### ISSUE TITLE
Improve Interoperability with other asynchronous libraries

### ISSUE THREAD

**Issue by illuhad** [illuhad/hipSYCL#181]
Based on a twitter discussion with @ax3l, we should investigate improving the interoperability of SYCL implementations with other asynchronous libraries (such as MPI or some async IO library).
This issue serves to track possible approaches and to discuss the specific requirements for users, so feedback is appreciated :)

At the moment I am thinking in particular of two new features:
1. Specify command group dependencies on some external events
2. Add callback mechanisms that trigger as soon as the SYCL implementation realizes that a task is complete (Note: depending on the task submission model implemented by the SYCL implementation, it may be possible that this callback will not be triggered right after a task completes; for example in a batched submission model the SYCL runtime may only notice that tasks are done once the entire batch is complete)

# 1. Specifying command group dependencies on some external events
I think external events could best be implemented on top of the explicit synchronization mechanism that is part of the Intel USM proposal which introduces `sycl::handler::depends_on(sycl::event evt)`, e.g.:

```c++
q.submit([&](cl::sycl::handler& cgh){
  cgh.depends_on(some_sycl_event);
  cgh.parallel_for(...);
});
```

In analogy, we could introduce an overload `sycl::handler::depends_on(sycl::external_event evt)`, with a new class `external_event`:
```c++
class external_event
{
public:
  external_event();
  // user can optionally provide a function that will be used by the SYCL implementation 
  // to test the state of the event. If this function is not provided, the event can only complete 
  // when   `signal_completion()` is called.
  // As soon as either the test_state function returns true, or the signal_completion() function
  // is called, the SYCL runtime will consider the event as complete.
  external_event(std::function<bool ()> test_state);

  void wait();

  // When called, signals to the SYCL runtime that this event has completed.
  void signal_completion();

  // ... plus remaining functions that are present in sycl::event for consistency
};
```

# 2. Add callback mechanisms 
TODO, I still need to think some more about that.

**Comment by ax3l** [comment]
MPI communication and long-running I/O routines are exactly what I face every day, yes!

Just a minor comment that probably works with this proposal: for the user-provided test function, there are some cases such as `MPI_Test` functions that can be queried an arbitrary amount of time *until* they return success for the first time (and they *must not* be queried again after that). One can probably express this somehow as a user-defined state (static?) when implementing this as a `test_state function`? Just want to mention this little odd use case.

**Comment by illuhad** [comment]
Thanks for the hint! I would expect that an implementation of `external_event` will typically contain a `bool` variable that specifies if the event has completed in order to support both the `signal_completion()` function as well as the user-provided test function. Under this assumption, the test function would probably not be called again anyway in typical implementations if the `bool` variable already indicates completion, which would be the behavior that you require.
In other words, since typical implementations of `external_event` would already work with such a restriction, I think we might as well just guarantee the user that the test function will not be called again after it returns success.

**Comment by ax3l** [comment]
That's a good constrain, thanks!

**Comment by illuhad** [comment]
Added a concept for callbacks.
